<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#2022년-1학기-금융시장의-빅데이터-분석-기말고사(담당교수:-박석진)" data-toc-modified-id="2022년-1학기-금융시장의-빅데이터-분석-기말고사(담당교수:-박석진)-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>2022년 1학기 금융시장의 빅데이터 분석 기말고사(담당교수: 박석진)</a></span></li><li><span><a href="#Regression-vs-Classification-(총-40점)" data-toc-modified-id="Regression-vs-Classification-(총-40점)-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Regression vs Classification (총 40점)</a></span></li></ul></div>

# 2022년 1학기 금융시장의 빅데이터 분석 기말고사(담당교수: 박석진)

# Regression vs Classification (총 40점)

- 이번 문제들에서는 Lasso example 노트북의 실현변동성(realized volatility) 데이터를 이용합니다.

- (1) (4점) 'data1.csv' 파일을 불러오고 맨 위, 맨 아래 행을 각각 5개씩 출력하세요.

In [1]:
import pandas as pd

df = pd.read_csv('./data1.csv', index_col='Unnamed: 0')
df.head(5)

Symbol  rk_twoscale    rk_th2  close_price  \
2000-01-03 00:00:00+00:00   .AEX     0.000103  0.000102       675.44   
2000-01-04 00:00:00+00:00   .AEX     0.000199  0.000201       642.25   
2000-01-05 00:00:00+00:00   .AEX     0.000325  0.000345       632.31   
2000-01-06 00:00:00+00:00   .AEX     0.000218  0.000221       624.21   
2000-01-07 00:00:00+00:00   .AEX     0.000126  0.000123       644.86   

                                rv5    rv5_ss  open_price      rv10  \
2000-01-03 00:00:00+00:00  0.000130  0.000130      675.67  0.000178   
2000-01-04 00:00:00+00:00  0.000201  0.000201      664.20  0.000261   
2000-01-05 00:00:00+00:00  0.000491  0.000491      633.37  0.000714   
2000-01-06 00:00:00+00:00  0.000225  0.000225      632.46  0.000182   
2000-01-07 00:00:00+00:00  0.000138  0.000138      628.93  0.000157   

                           rk_parzen       rsv     medrv     bv_ss    rsv_ss  \
2000-01-03 00:00:00+00:00   0.000179  0.000046  0.000050  0.000100  0.000046   
2000-01-04 00:00:00+00:00   0.000423  0.000147  0.000075  0.000207  0.000147   
2000-01-05 00:00:00+00:00   0.000324  0.000328  0.000166  0.000361  0.000328   
2000-01-06 00:00:00+00:00   0.000219  0.000116  0.000152  0.000258  0.000116   
2000-01-07 00:00:00+00:00   0.000155  0.000048  0.000039  0.000130  0.000048   

                           open_time   rv10_ss    nobs        bv  \
2000-01-03 00:00:00+00:00    90101.0  0.000178  1795.0  0.000100   
2000-01-04 00:00:00+00:00    90416.0  0.000261  1785.0  0.000207   
2000-01-05 00:00:00+00:00    90016.0  0.000714  1801.0  0.000361   
2000-01-06 00:00:00+00:00    90016.0  0.000182  1799.0  0.000258   
2000-01-07 00:00:00+00:00    90046.0  0.000157  1798.0  0.000130   

                           open_to_close  close_time  
2000-01-03 00:00:00+00:00      -0.000340    163015.0  
2000-01-04 00:00:00+00:00      -0.033606    163016.0  
2000-01-05 00:00:00+00:00      -0.001675    163016.0  
2000-01-06 00:00:00+00:00      -0.013130    163002.0  
2000-01-07 00:00:00+00:00       0.025013    163016.0

In [2]:
df.tail(5)

Symbol  rk_twoscale    rk_th2  close_price  \
2021-11-15 00:00:00+00:00  .STOXX50E     0.000012  0.000012      4381.38   
2021-11-16 00:00:00+00:00  .STOXX50E     0.000015  0.000015      4402.04   
2021-11-17 00:00:00+00:00  .STOXX50E     0.000009  0.000009      4402.69   
2021-11-18 00:00:00+00:00  .STOXX50E     0.000014  0.000014      4383.91   
2021-11-19 00:00:00+00:00  .STOXX50E     0.000103  0.000105      4361.27   

                                rv5    rv5_ss  open_price      rv10  \
2021-11-15 00:00:00+00:00  0.000015  0.000015     4364.52  0.000012   
2021-11-16 00:00:00+00:00  0.000011  0.000011     4393.80  0.000010   
2021-11-17 00:00:00+00:00  0.000009  0.000009     4399.25  0.000007   
2021-11-18 00:00:00+00:00  0.000011  0.000011     4400.73  0.000011   
2021-11-19 00:00:00+00:00  0.000094  0.000094     4385.66  0.000125   

                           rk_parzen       rsv     medrv     bv_ss    rsv_ss  \
2021-11-15 00:00:00+00:00   0.000008  0.000005  0.000004  0.000013  0.000005   
2021-11-16 00:00:00+00:00   0.000006  0.000004  0.000005  0.000012  0.000004   
2021-11-17 00:00:00+00:00   0.000005  0.000005  0.000005  0.000008  0.000005   
2021-11-18 00:00:00+00:00   0.000017  0.000007  0.000004  0.000010  0.000007   
2021-11-19 00:00:00+00:00   0.000117  0.000050  0.000047  0.000086  0.000050   

                           open_time   rv10_ss    nobs        bv  \
2021-11-15 00:00:00+00:00    90015.0  0.000012  4116.0  0.000013   
2021-11-16 00:00:00+00:00    90015.0  0.000010  4210.0  0.000012   
2021-11-17 00:00:00+00:00    90015.0  0.000007  4247.0  0.000008   
2021-11-18 00:00:00+00:00    90015.0  0.000011  4153.0  0.000010   
2021-11-19 00:00:00+00:00    90015.0  0.000125  4190.0  0.000086   

                           open_to_close  close_time  
2021-11-15 00:00:00+00:00       0.003856    173004.0  
2021-11-16 00:00:00+00:00       0.001874    173002.0  
2021-11-17 00:00:00+00:00       0.000782    173015.0  
2021-11-18 00:00:00+00:00      -0.003829    173002.0  
2021-11-19 00:00:00+00:00      -0.005577    173003.0

- (2) (4점) 데이터 중 다우존스 지수(symbol=DJI)에 해당하는 데이터 중 'rv5' 만을 따로 뽑아내서 새로운 dataframe을 구성하고, rv5의 과거값을 lag1 부터 lag 20까지를 각각 개별적인 column으로 추가하세요.

In [3]:
dow = df[df['Symbol']=='.DJI'][['rv5']]
dow

rv5
2000-01-03 00:00:00+00:00  0.000124
2000-01-04 00:00:00+00:00  0.000211
2000-01-05 00:00:00+00:00  0.000211
2000-01-06 00:00:00+00:00  0.000114
2000-01-07 00:00:00+00:00  0.000122
...                             ...
2021-11-15 00:00:00+00:00  0.000019
2021-11-16 00:00:00+00:00  0.000022
2021-11-17 00:00:00+00:00  0.000026
2021-11-18 00:00:00+00:00  0.000020
2021-11-19 00:00:00+00:00  0.000055

[5484 rows x 1 columns]

In [4]:
for i in range(1, 21):
    dow['rv5_'+str(i)] = dow['rv5'].shift(i)
dow = dow.dropna()
dow

rv5     rv5_1     rv5_2     rv5_3     rv5_4  \
2000-02-01 00:00:00+00:00  0.000125  0.000188  0.000291  0.000174  0.000087   
2000-02-02 00:00:00+00:00  0.000146  0.000125  0.000188  0.000291  0.000174   
2000-02-03 00:00:00+00:00  0.000144  0.000146  0.000125  0.000188  0.000291   
2000-02-04 00:00:00+00:00  0.000101  0.000144  0.000146  0.000125  0.000188   
2000-02-07 00:00:00+00:00  0.000092  0.000101  0.000144  0.000146  0.000125   
...                             ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  0.000019  0.000017  0.000010  0.000025  0.000023   
2021-11-16 00:00:00+00:00  0.000022  0.000019  0.000017  0.000010  0.000025   
2021-11-17 00:00:00+00:00  0.000026  0.000022  0.000019  0.000017  0.000010   
2021-11-18 00:00:00+00:00  0.000020  0.000026  0.000022  0.000019  0.000017   
2021-11-19 00:00:00+00:00  0.000055  0.000020  0.000026  0.000022  0.000019   

                              rv5_5     rv5_6     rv5_7     rv5_8     rv5_9  \
2000-02-01 00:00:00+00:00  0.000164  0.000277  0.000073  0.000160  0.000063   
2000-02-02 00:00:00+00:00  0.000087  0.000164  0.000277  0.000073  0.000160   
2000-02-03 00:00:00+00:00  0.000174  0.000087  0.000164  0.000277  0.000073   
2000-02-04 00:00:00+00:00  0.000291  0.000174  0.000087  0.000164  0.000277   
2000-02-07 00:00:00+00:00  0.000188  0.000291  0.000174  0.000087  0.000164   
...                             ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  0.000025  0.000031  0.000016  0.000025  0.000012   
2021-11-16 00:00:00+00:00  0.000023  0.000025  0.000031  0.000016  0.000025   
2021-11-17 00:00:00+00:00  0.000025  0.000023  0.000025  0.000031  0.000016   
2021-11-18 00:00:00+00:00  0.000010  0.000025  0.000023  0.000025  0.000031   
2021-11-19 00:00:00+00:00  0.000017  0.000010  0.000025  0.000023  0.000025   

                           ...    rv5_11    rv5_12    rv5_13    rv5_14  \
2000-02-01 00:00:00+00:00  ...  0.000085  0.000067  0.000101  0.000068   
2000-02-02 00:00:00+00:00  ...  0.000117  0.000085  0.000067  0.000101   
2000-02-03 00:00:00+00:00  ...  0.000063  0.000117  0.000085  0.000067   
2000-02-04 00:00:00+00:00  ...  0.000160  0.000063  0.000117  0.000085   
2000-02-07 00:00:00+00:00  ...  0.000073  0.000160  0.000063  0.000117   
...                        ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  ...  0.000019  0.000025  0.000029  0.000017   
2021-11-16 00:00:00+00:00  ...  0.000043  0.000019  0.000025  0.000029   
2021-11-17 00:00:00+00:00  ...  0.000012  0.000043  0.000019  0.000025   
2021-11-18 00:00:00+00:00  ...  0.000025  0.000012  0.000043  0.000019   
2021-11-19 00:00:00+00:00  ...  0.000016  0.000025  0.000012  0.000043   

                             rv5_15    rv5_16    rv5_17    rv5_18    rv5_19  \
2000-02-01 00:00:00+00:00  0.000072  0.000122  0.000114  0.000211  0.000211   
2000-02-02 00:00:00+00:00  0.000068  0.000072  0.000122  0.000114  0.000211   
2000-02-03 00:00:00+00:00  0.000101  0.000068  0.000072  0.000122  0.000114   
2000-02-04 00:00:00+00:00  0.000067  0.000101  0.000068  0.000072  0.000122   
2000-02-07 00:00:00+00:00  0.000085  0.000067  0.000101  0.000068  0.000072   
...                             ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  0.000013  0.000027  0.000012  0.000012  0.000017   
2021-11-16 00:00:00+00:00  0.000017  0.000013  0.000027  0.000012  0.000012   
2021-11-17 00:00:00+00:00  0.000029  0.000017  0.000013  0.000027  0.000012   
2021-11-18 00:00:00+00:00  0.000025  0.000029  0.000017  0.000013  0.000027   
2021-11-19 00:00:00+00:00  0.000019  0.000025  0.000029  0.000017  0.000013   

                             rv5_20  
2000-02-01 00:00:00+00:00  0.000124  
2000-02-02 00:00:00+00:00  0.000211  
2000-02-03 00:00:00+00:00  0.000211  
2000-02-04 00:00:00+00:00  0.000114  
2000-02-07 00:00:00+00:00  0.000122  
...                             ...  
2021-11-15 00:00:00+00:00  0.0

- (3) (5점) 시계열 자료가 안정적인지 검증하세요.

In [5]:
# stationary test
from statsmodels.tsa.stattools import adfuller

print(adfuller(dow.rv5, regression='n')[1])
print(adfuller(dow.rv5, regression='c')[1])
print(adfuller(dow.rv5, regression='ct')[1])

9.292638493182099e-11
1.6168079129582573e-12
4.1788023974297335e-11


어떤 regression이든 p value < 0.05이므로 귀무가설을 기각한다. 따라서 dow 지수의 rv5 자료는 stationary한 시계열 자료이다.

- (4)  (4점) Dataframe의 rv5를 y변수로, rv5를 제외한 모든 lag 변수들을 X 변수로 설정하고, sklearn의 train_test_split을 이용해 train set과 test set을 구성하세요(이때, train_test_split에서 test size는 0.3, random state는 42로 설정).

In [6]:
from sklearn.model_selection import train_test_split
y = dow['rv5']
X = dow.drop('rv5', axis=1)
X_train_rd, X_test_rd, y_train_rd, y_test_rd = train_test_split(X, y, test_size=0.3,
                                                    random_state=42)

- (5) (5점) sklearn의 Ridge regression을 train set으로 훈련하고, test set에 대해 예측을 진행하십시오. Test set에 대한 LinearRegression 모형의 $R^2$ score를 도출하십시오. 위의 모형이 rv5에 대해 예측력을 가진다고 말할 수 있습니까? (단, hyperparameter는 따로 tuning하지 않고 alpha만 0.001로 설정합니다.)

In [7]:
from sklearn.linear_model import Ridge
from sklearn.metrics import r2_score

ridge_reg = Ridge(alpha=0.001)
ridge_reg.fit(X_train_rd, y_train_rd)
y_pred_rd = ridge_reg.predict(X_test_rd)
print('r2 score =', r2_score(y_test_rd, y_pred_rd))

r2 score = 0.3097370290254279


r2 score는 1에 가까울 수록 성능이 좋고 0에 가까울 수록 성능이 나쁘다는 의미인데, 일반적으로 r square가 0.3 이상이면 종속변수에 대한 예측력을 가진다고 본다. 즉 이 모델은 rv5에 대한 예측력을 가진다고 할 수 있다.

- (6) (6점) rv5가 직전 관측치보다 상승했으면 1, 감소하거나 변하지 않았으면 0을 부여하고 이를 y변수로 설정합니다. sklearn의 Logistic regression을 train set으로 훈련하고, test set에 대한 Logistic regression 모형의 classification report를 도출하십시오(단, hyperparameter는 따로 tuning하지 않고 default 값 그대로 사용합니다). 이 결과를 바탕으로, logistic regression 모형이 rv5의 움직임에 대해 예측력을 가진다고 말할 수 있습니까?

In [8]:
dow['target'] = 0

for i in range(1, len(dow)):
    if dow.iloc[i, 0] >  dow.iloc[i-1, 0]:
        dow.iloc[i, -1] = 1

C:\Users\LOVE.DESKTOP-5CS2DDB\AppData\Local\Temp\ipykernel_3848\817086752.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dow['target'] = 0


In [9]:
dow.drop('2000-02-01 00:00:00+00:00', axis=0)

rv5     rv5_1     rv5_2     rv5_3     rv5_4  \
2000-02-02 00:00:00+00:00  0.000146  0.000125  0.000188  0.000291  0.000174   
2000-02-03 00:00:00+00:00  0.000144  0.000146  0.000125  0.000188  0.000291   
2000-02-04 00:00:00+00:00  0.000101  0.000144  0.000146  0.000125  0.000188   
2000-02-07 00:00:00+00:00  0.000092  0.000101  0.000144  0.000146  0.000125   
2000-02-08 00:00:00+00:00  0.000083  0.000092  0.000101  0.000144  0.000146   
...                             ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  0.000019  0.000017  0.000010  0.000025  0.000023   
2021-11-16 00:00:00+00:00  0.000022  0.000019  0.000017  0.000010  0.000025   
2021-11-17 00:00:00+00:00  0.000026  0.000022  0.000019  0.000017  0.000010   
2021-11-18 00:00:00+00:00  0.000020  0.000026  0.000022  0.000019  0.000017   
2021-11-19 00:00:00+00:00  0.000055  0.000020  0.000026  0.000022  0.000019   

                              rv5_5     rv5_6     rv5_7     rv5_8     rv5_9  \
2000-02-02 00:00:00+00:00  0.000087  0.000164  0.000277  0.000073  0.000160   
2000-02-03 00:00:00+00:00  0.000174  0.000087  0.000164  0.000277  0.000073   
2000-02-04 00:00:00+00:00  0.000291  0.000174  0.000087  0.000164  0.000277   
2000-02-07 00:00:00+00:00  0.000188  0.000291  0.000174  0.000087  0.000164   
2000-02-08 00:00:00+00:00  0.000125  0.000188  0.000291  0.000174  0.000087   
...                             ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  0.000025  0.000031  0.000016  0.000025  0.000012   
2021-11-16 00:00:00+00:00  0.000023  0.000025  0.000031  0.000016  0.000025   
2021-11-17 00:00:00+00:00  0.000025  0.000023  0.000025  0.000031  0.000016   
2021-11-18 00:00:00+00:00  0.000010  0.000025  0.000023  0.000025  0.000031   
2021-11-19 00:00:00+00:00  0.000017  0.000010  0.000025  0.000023  0.000025   

                           ...    rv5_12    rv5_13    rv5_14    rv5_15  \
2000-02-02 00:00:00+00:00  ...  0.000085  0.000067  0.000101  0.000068   
2000-02-03 00:00:00+00:00  ...  0.000117  0.000085  0.000067  0.000101   
2000-02-04 00:00:00+00:00  ...  0.000063  0.000117  0.000085  0.000067   
2000-02-07 00:00:00+00:00  ...  0.000160  0.000063  0.000117  0.000085   
2000-02-08 00:00:00+00:00  ...  0.000073  0.000160  0.000063  0.000117   
...                        ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  ...  0.000025  0.000029  0.000017  0.000013   
2021-11-16 00:00:00+00:00  ...  0.000019  0.000025  0.000029  0.000017   
2021-11-17 00:00:00+00:00  ...  0.000043  0.000019  0.000025  0.000029   
2021-11-18 00:00:00+00:00  ...  0.000012  0.000043  0.000019  0.000025   
2021-11-19 00:00:00+00:00  ...  0.000025  0.000012  0.000043  0.000019   

                             rv5_16    rv5_17    rv5_18    rv5_19    rv5_20  \
2000-02-02 00:00:00+00:00  0.000072  0.000122  0.000114  0.000211  0.000211   
2000-02-03 00:00:00+00:00  0.000068  0.000072  0.000122  0.000114  0.000211   
2000-02-04 00:00:00+00:00  0.000101  0.000068  0.000072  0.000122  0.000114   
2000-02-07 00:00:00+00:00  0.000067  0.000101  0.000068  0.000072  0.000122   
2000-02-08 00:00:00+00:00  0.000085  0.000067  0.000101  0.000068  0.000072   
...                             ...       ...       ...       ...       ...   
2021-11-15 00:00:00+00:00  0.000027  0.000012  0.000012  0.000017  0.000036   
2021-11-16 00:00:00+00:00  0.000013  0.000027  0.000012  0.000012  0.000017   
2021-11-17 00:00:00+00:00  0.000017  0.000013  0.000027  0.000012  0.000012   
2021-11-18 00:00:00+00:00  0.000029  0.000017  0.000013  0.000027  0.000012   
2021-11-19 00:00:00+00:00  0.000025  0.000029  0.000017  0.000013  0.000027   

                           target  
2000-02-02 00:00:00+00:00       1  
2000-02-03 00:00:00+00:00       0  
2000-02-04 00:00:00+00:00       0  
2000-02-07 00:00:00+00:00       0  
2000-02-08 00:00:00+00:00       0  
...                           ...  
2021-11-15 00:00:00+00:00       1  
2021-11-

첫번째 행은 이전의 관측치가 없으므로 삭제하였다.

In [10]:
from sklearn.model_selection import train_test_split
y = dow['target']
X = dow.drop('target', axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,
                                                    random_state=42)

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score, confusion_matrix, classification_report


lg = LogisticRegression()
lg.fit(X_train,y_train)
predicted = lg.predict(X_test)
print(classification_report(y_test, predicted))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       824
           1       0.50      1.00      0.66       816

    accuracy                           0.50      1640
   macro avg       0.25      0.50      0.33      1640
weighted avg       0.25      0.50      0.33      1640



C:\Users\LOVE.DESKTOP-5CS2DDB\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1327: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\LOVE.DESKTOP-5CS2DDB\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1327: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\LOVE.DESKTOP-5CS2DDB\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1327: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


accuracy만 확인하면 정확도가 0.5니 설명력이 있다고 착각할 수 있다. 그러나 클래스별 f1-score를 보면, 클래스 1에 대해서는 0.66인데 반해 클래스 0에 대해서는 0.00이다. 이를 통해 rv5가 하락한 케이스에 대해서는 예측력이 없는 모델이라는 것을 알 수 있다. 또한 클래스 1의 세부 지표를 살펴보면, recall은 1.00인데 반해 precision은 0.50이다. 이 말은 target이 1일 때 모델이 전부 1을 예측했으나, 모델이 1을 예측했을 때 정답이 1인 경우는 절반밖에 되지 않았다는 것이다. 결국 이 모델은 대다수의 test 데이터에 대해 1을 예측하는 잘못된 모델이라는 것을 추측할 수 있다.

In [12]:
pd.DataFrame({'pd':predicted}).value_counts()

pd
1     1640
dtype: int64

실제로 예측치를 확인해보면, 전부 클래스 1뿐이다. 즉 이 모델은 rv5의 움직임에 대해서 예측력이 없다.

- (7) (6점) 다음 날의 rv5가 오늘의 rv5와 같을 것이라는 모형(즉, $\hat{rv}_{t+1} = rv_t$)의 $R^2$ score를 (5)의 test set을 이용해 도출하고, 이 결과를 바탕으로 (5)와 (6)의 결과를 해석하세요.

In [13]:
X_test_rd = X_test_rd.sort_index()
y_test_rd = y_test_rd.sort_index()
y_predict = y_test_rd.copy()

for row in range(1, len(y_predict)):
    y_predict[row] = y_predict[row-1]
print('r2 score =', r2_score(y_test_rd, y_predict))

r2 score = -0.024229924204585362


In [14]:
y_predict

2000-02-11 00:00:00+00:00    0.000148
2000-02-17 00:00:00+00:00    0.000148
2000-02-23 00:00:00+00:00    0.000148
2000-02-25 00:00:00+00:00    0.000148
2000-02-29 00:00:00+00:00    0.000148
                               ...   
2021-10-29 00:00:00+01:00    0.000148
2021-11-02 00:00:00+00:00    0.000148
2021-11-10 00:00:00+00:00    0.000148
2021-11-11 00:00:00+00:00    0.000148
2021-11-19 00:00:00+00:00    0.000148
Name: rv5, Length: 1640, dtype: float64

다음날의 rv5가 오늘의 rv5와 같으므로, 첫번째 날의 rv5가 모든 예측치가 되도록 할당해주었다. 그리고 test set과 r square score를 구해보자 음수가 나온다. 이는 종속변수의 일괄 평균값으로 예측하는 것보다 낮은 성능을 보인다는 의미이므로 엉터리 모델이라고 할 수 있다.  
이에 반해 (5)의 경우는 양수이며 30%의 설명력이 있으므로 제대로 된 모델이다. 또한 (6)의 경우는 이 엉터리 모델과 유사하게 모든 예측치가 1이었지만 accuracy는 0.5 수준으로 나왔다. 그러므로 모델링을 할 때는 단순 score만 보고 모델의 예측력을 판단하는 것이 아니라 세부적인 지표까지 참고해야 한다는 것을 알 수 있다.

- (8) (6점) (6)의 분석을 SVM과 Random Forest로 다시 진행하고(단, hyperparameter는 따로 tuning하지 않고 default 값 그대로 사용합니다), test set에 대한 각 모형의 classification report를 도출합니다.

In [15]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.93      0.91      0.92       824
           1       0.91      0.93      0.92       816

    accuracy                           0.92      1640
   macro avg       0.92      0.92      0.92      1640
weighted avg       0.92      0.92      0.92      1640



전체적인 accurace score가 0.93으로 굉장히 높다. 세부적으로 살펴보면, 클래스 0과 1에 대해 모두 0.93의 높은 f1-score를 가진다. 또한 precision과 recall도 0.92~0.94 수준으로 높은 편이므로 이 모델은 rv5의 움직임에 대해 예측력이 높은 모델이라 할 수 있다.

In [16]:
from sklearn.svm import SVC, LinearSVC
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score, confusion_matrix, classification_report
from sklearn.model_selection import train_test_split, KFold, cross_val_score


svc = SVC()
svc.fit(X_train,y_train)
predicted = svc.predict(X_test)
print(classification_report(y_test, predicted))

              precision    recall  f1-score   support

           0       0.91      0.69      0.79       824
           1       0.75      0.93      0.83       816

    accuracy                           0.81      1640
   macro avg       0.83      0.81      0.81      1640
weighted avg       0.83      0.81      0.81      1640



전체 accuracy score가 0.81 수준으로 앞선 RandomForest보다는 낮다. 그러나 각 클래스별 f1-score가 0.79에서 0.83수준이고, precision과 recall도 약간의 차이는 있으나 크게 편향된 수준은 아니다. 그러므로 이 모델은 rv5의 움직임에 대해서 예측력이 높은 모델이라고 할 수 있다.